src/core/expression/qgsexpressioncontextutils.cpp

```diff
@@ -1024,13 +1024,24 @@ class CurrentVertexZValueExpressionFunction: public QgsScopedExpressionFunction
 
       int vertexIndex = context->variable( QStringLiteral( "_mesh_vertex_index" ) ).toInt();
 
-      QgsMeshLayer *layer = QgsExpressionUtils::getMeshLayer( context->variable( QStringLiteral( "layer" ) ), parent );
-      if ( !layer || !layer->nativeMesh() || layer->nativeMesh()->vertexCount() <= vertexIndex )
-        return QVariant();
+      auto lambdaFunc = [vertexIndex]( QgsMapLayer * layer )->QVariant
+      {
+        QgsMeshLayer *meshLayer = qobject_cast<QgsMeshLayer *>( layer );
+        if ( !layer || !meshLayer->nativeMesh() || meshLayer->nativeMesh()->vertexCount() <= vertexIndex )
```

Comment:
```suggestion
        if ( !meshLayer || !meshLayer->nativeMesh() || meshLayer->nativeMesh()->vertexCount() <= vertexIndex )
```
in other lambdas too